packages/access/lib/grants.js

```diff
@@ -0,0 +1,202 @@
+    }
+
+    const result = await constraints[name].isGrantable(
+      { settings, grantee, email, campaign },
+      { pgdb }
+    )
+
+    debug('isGrantable', {
+      name: campaign.name,
+      constraint: name,
+      settings,
+      result
+    })
+
+    if (!result) {
+      granted = false
+      break
+    }
+  }
+
+  debug('granted', granted)
+
+  return granted
+}
+
+const grant = async (grantee, campaignId, email, t, pgdb) => {
+  if (!validator.isEmail(email)) {
+    throw new Error(`Email address "${email}" is invalid`)
+  }
+
+  const campaign = await campaignsLib.findOne(campaignId, pgdb)
+
+  if (campaign === undefined) {
+    throw new Error(`campaign "${campaignId}" not found`)
+  }
+
+  if (!await isGrantable(grantee, campaign, email, pgdb)) {
+    throw new Error('Unable to grant membership')
+  }
+
+  const endAt = moment()
+
+  Object.keys(campaign.periodInterval).forEach(key => {
+    endAt.add(campaign.periodInterval[key], key)
+  })
+
+  const grant = await pgdb.public.accessGrants.insertAndGet({
+    granteeUserId: grantee.id,
+    email,
+    accessCampaignId: campaign.id,
+    beginAt: moment(),
+    endAt
+  })
+
+  eventsLib.log(grant, 'grant', pgdb)
+
+  debug('grant, row inserted', { grant })
+
+  await Promise.all([
+    mailLib.sendRecipientOnboarding(grantee, campaign, grant, t, pgdb),
+    mailLib.sendGranteeConfirmation(grantee, campaign, grant, t, pgdb)
+  ])
+
+  return grant
+}
+
+const revoke = async (id, grantee, t, pgdb, mail) => {
+  const grant = await pgdb.public.accessGrants.findOne({ id })
+
+  return renderInvalid(grant, 'revoked', pgdb, mail)
+}
+
+const findByGrantee = async (grantee, campaign, pgdb) => {
+  const grants =
+    await pgdb.public.accessGrants.find({
+      granteeUserId: grantee.id,
+      accessCampaignId: campaign.id,
+      'beginAt <=': moment(),
+      'endAt >': moment(),
+      invalidatedAt: null
+    })
+
+  return grants
+}
+
+const findByRecipient = async (recipient, pgdb) => {
+  const grants =
+    await pgdb.public.accessGrants.find({
+      recipientUserId: recipient.id,
+      'beginAt <=': moment(),
+      'endAt >': moment(),
+      invalidatedAt: null
+    })
+
+  return grants
+}
+
+const findUnassigned = async (pgdb) => pgdb.public.accessGrants.find({
+  recipientUserId: null,
+  'beginAt <=': moment(),
+  'endAt >': moment(),
+  invalidatedAt: null
+})
+
+const findExpired = async (pgdb) => pgdb.public.accessGrants.find({
+  'endAt <': moment(),
+  invalidatedAt: null
+})
+
+const setRecipient = async (grant, recipient, pgdb) => {
+  const result = await pgdb.public.accessGrants.update(
+    { id: grant.id },
+    { recipientUserId: recipient.id, updatedAt: moment() }
+  )
+
+  debug('setRecipient', {
+    id: grant.id,
+    recipientUserId: recipient.id,
+    updatedAt: moment(),
+    result
+  })
+
+  return result > 0
+}
+
+const renderInvalid = async (grant, reason, pgdb, mail) => {
+  const result = await pgdb.public.accessGrants.update(
```

Comment:
Even though the name is perfectly descriptive, in our codebase I'd be hesitant to call any method renderSomething that is not actually rendering something, because that term is strongly connected to React et al.

Reply:
`invalidate` would be a good alternative I think